[V1] Support MP Executor for multi node distributed inference

luccafong opened the pull request:
## Purpose

Support MP Executor for multi node distributed inference when no ray setup. (Compatible with DP hybrid/internal/external lb)


- How to use:
`vllm serve "model_name" -tp <TP_Size> -dp <DP_Size> -pp <PP_Size> --nnodes <# nodes> --node-rank <rank of node>  --master-addr <leader_host_ip>  [--master-port <port>] [--headless (if not exposing API endpoint)>] [--data-parallel-external-lb (for external lb) or --data-parallel-hybrid-lb]`
See concrete examples in test plans

**Note for DP Compatibility:**
- We don't implement new DP coordinator/backend in this PR, we reuse what we already have.
- This PR support auto inference of `--data-parallel-rank` `--data-parallel-local-size` `--data-parallel-start-rank` based on `-distributed-node-rank` for all 3 dp lb modes, and it works the same with using them explicitly before.
- For engines that do not expose api server (e.g. leader or external dp engine), we need to add `--headless` (it will auto start a headless DP engine if it  is leader(driver) of a DP group; or start a headless Executor Instance if it is non-driver workers)
- This is backward compatible with existing interface for other data parallel.

Architecture change

<img width="2948" height="1512" alt="image" src="https://github.com/user-attachments/assets/77a3735b-dd24-4e04-a967-b6578e34325b" />


## Test Plan

### TP=4 (2 Instances )

CUDA_VISIBLE_DEVICES=0,1 vllm serve "Qwen/Qwen3-1.7B" -tp=4 --max-model-len=32768 --nnodes 2 --node-rank 0 --master-addr 127.0.0.1 --port 8000 > /tmp/test_rank0.log 2>&1 &

CUDA_VISIBLE_DEVICES=2,3 vllm serve "Qwen/Qwen3-1.7B"  -tp=4 --max-model-len=32768 --nnodes 2 --node-rank 1 --master-addr 127.0.0.1 --headless > /tmp/test_rank1.log 2>&1 &


### PP=2 x TP=2  (2 Instances)

CUDA_VISIBLE_DEVICES=0,1 vllm serve "Qwen/Qwen3-1.7B" -pp=2 -tp=2 --max-model-len=32768 --nnodes 2 --node-rank 0 --master-addr 127.0.0.1 --port 8000 > /tmp/test_rank0.log 2>&1 &

CUDA_VISIBLE_DEVICES=2,3 vllm serve "Qwen/Qwen3-1.7B" -pp=2 -tp=2 --max-model-len=32768 --nnodes 2 --node-rank 1 --master-addr 127.0.0.1 --headless > /tmp/test_rank1.log 2>&1 &



### DP 2 (External) * TP 4 (2 in node, 2 across node) (4 Instances)


CUDA_VISIBLE_DEVICES=0,1 vllm serve "Qwen/Qwen3-1.7B" -dp=2 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank 0 --master-addr 127.0.0.1 --port 8000 --data-parallel-external-lb > /tmp/test_dp0.log 2>&1 &

CUDA_VISIBLE_DEVICES=2,3 vllm serve "Qwen/Qwen3-1.7B" -dp=2 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank 1 --master-addr 127.0.0.1  --headless > /tmp/test_dp1.log 2>&1 &

CUDA_VISIBLE_DEVICES=4,5 vllm serve "Qwen/Qwen3-1.7B" -dp=2 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank  2 --master-addr 127.0.0.1   --port 8001  --data-parallel-external-lb  > /tmp/test_dp2.log 2>&1 &

CUDA_VISIBLE_DEVICES=6,7 vllm serve "Qwen/Qwen3-1.7B" -dp=2 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank  3 --master-addr 127.0.0.1 --headless  > /tmp/test_dp3.log 2>&1 &

### DP 2 (Internal with headless) * TP 4 (2 in node, 2 across node) (4 Instances)

CUDA_VISIBLE_DEVICES=0,1 vllm serve "Qwen/Qwen3-1.7B" -dp=2 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank 0 --master-addr 127.0.0.1 --port 8000 > /tmp/test_dp0.log 2>&1 &

CUDA_VISIBLE_DEVICES=2,3 vllm serve "Qwen/Qwen3-1.7B" -dp=2 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank 1 --master-addr 127.0.0.1 --headless > /tmp/test_dp1.log 2>&1 &

CUDA_VISIBLE_DEVICES=4,5 vllm serve "Qwen/Qwen3-1.7B" -dp=2 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank  2 --master-addr 127.0.0.1 --headless  > /tmp/test_dp2.log 2>&1 &

CUDA_VISIBLE_DEVICES=6,7 vllm serve "Qwen/Qwen3-1.7B" -dp=2 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank  3 --master-addr 127.0.0.1 --headless > /tmp/test_dp3.log 2>&1 &

### DP * 4 (Internal) * TP 2 (inner node) (4 Instances)

CUDA_VISIBLE_DEVICES=0,1 vllm serve "Qwen/Qwen3-1.7B" -dp=4 -tp=2 --max-model-len=32768 --nnodes 4 --node-rank 0 --master-addr 127.0.0.1 --port 8000 > /tmp/test_dp0.log 2>&1 &

CUDA_VISIBLE_DEVICES=2,3 vllm serve "Qwen/Qwen3-1.7B" -dp=4 -tp=4 --max-model-len=32768 --nnodes 4 --node-rank 1 --master-addr 127.0.0.1 --headless > /tmp/test_dp1.log 2>&1 &

CUDA_VISIBLE_DEVICES=4,5 vllm serve "Qwen/Qwen3-1.7B" -dp=4 -tp=2 --max-model-len=32768 --nnodes 4 --node-rank  2 --master-addr 127.0.0.1 --headless  > /tmp/test_dp2.log 2>&1 &

CUDA_VISIBLE_DEVICES=6,7 vllm serve "Qwen/Qwen3-1.7B" -dp=4 -tp=2 --max-model-len=32768 --nnodes 4 --node-rank  3 --master-addr 127.0.0.1 --headless > /tmp/test_dp3.log 2>&1 &

### DP * 4 (External) * TP 2 (inner node) (4 Instances)

CUDA_VISIBLE_DEVICES=0,1 vllm serve "Qwen/Qwen3-1.7B" -dp=4 -tp=2 --max-model-len=32768 --nnodes 4 --node-rank 0 --master-addr 127.0.0.1 --port 8000 --data-parallel-external-lb > /tmp/test_dp0.log 2>&1 &

CUDA_VISIBLE_DEVICES=2,3 vllm serve "Qwen/Qwen3-1.7B" -dp=4 -tp=2 --max-model-len=32768 --nnodes 4 --node-rank 1 --master-addr 127.0.0.1  --port 8001 --data-parallel-external-lb > /tmp/test_dp1.log 2>&1 &

CUDA_VISIBLE_DEVICES=4,5 vllm serve "Qwen/Qwen3-1.7B" -dp=4 -tp=2 --max-model-len=32768 --nnodes 4 --node-rank  2 --master-addr 127.0.0.1   --port 8002  --data-parallel-external-lb  > /tmp/test_dp2.log 2>&1 &

CUDA_VISIBLE_DEVICES=6,7 vllm serve "Qwen/Qwen3-1.7B" -dp=4 -tp=2 --max-model-len=32768 --nnodes 4 --node-rank  3 --master-addr 127.0.0.1 --port 8003  --data-parallel-external-lb > /tmp/test_dp3.log 2>&1 &

### DP * 2(external)  * 2(internal) Hybrid. * TP * 2 (2 Instances)

CUDA_VISIBLE_DEVICES=0,1,2,3  vllm serve "Qwen/Qwen3-1.7B" -dp=4  -tp=2 --max-model-len=32768  --nnodes 2  --node-rank 0  --master-addr  127.0.0.1 --data-parallel-hybrid-lb  > /tmp/test_dp0.log 2>&1 &

CUDA_VISIBLE_DEVICES=4,5,6,7  vllm serve "Qwen/Qwen3-1.7B" -dp=4  -tp=2 --max-model-len=32768  --nnodes 2  --node-rank 1  --master-addr  127.0.0.1    --data-parallel-hybrid-lb  --port 8002  > /tmp/test_dp1.log 2>&1 &

### DP * 2  (internal) * PP * 2  * TP * 2(across nodes)   (8 Instances)

```
for i in {0..7}; do
     HEADLESS=""
     if [ $i -gt 0 ]; then HEADLESS="--headless";fi
    CUDA_VISIBLE_DEVICES=$i vllm serve Qwen/Qwen3-1.7B -dp=2  -pp=2 -tp=2 --max-model-len=32768 -n 8 -r  $i --master-addr 127.0.0.1  --port 8000 $HEADLESS > /tmp/test_mp${i}.log 2>&1 &
done
```

## Test Result

Eval tested on all above combinations and on par with baseline, below are some examples.

### TP Eval

```
lm_eval --model local-completions --tasks gsm8k     --model_args model=Qwen/Qwen3-1.7B,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=200,max_retries=3,tokenized_requests=False     --limit 1000
```


|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.681|±  |0.0147|
|     |       |strict-match    |     5|exact_match|↑  |0.678|±  |0.0148|



Baseline
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.681|±  |0.0147|
|     |       |strict-match    |     5|exact_match|↑  |0.679|±  |0.0148|



### TP Perf

```
vllm bench serve --model Qwen/Qwen3-1.7B  --port 8003  --dataset-name random  --ignore-eos  --num-prompts 512  --request-rate inf  --random-input-len 1024  --random-output-len 128  --max-concurrency 128
```

```
============ Serving Benchmark Result ============
Successful requests:                     512
Maximum request concurrency:             128
Benchmark duration (s):                  5.22
Total input tokens:                      522598
Total generated tokens:                  65536
Request throughput (req/s):              98.05
Output token throughput (tok/s):         12550.62
Total Token throughput (tok/s):          112631.95
---------------Time to First Token----------------
Mean TTFT (ms):                          225.67
Median TTFT (ms):                        235.68
P99 TTFT (ms):                           517.28
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          8.39
Median TPOT (ms):                        8.42
P99 TPOT (ms):                           9.68
---------------Inter-token Latency----------------
```

Baseline

```
============ Serving Benchmark Result ============
Successful requests:                     512
Maximum request concurrency:             128
Benchmark duration (s):                  5.17
Total input tokens:                      522598
Total generated tokens:                  65536
Request throughput (req/s):              99.08
Output token throughput (tok/s):         12682.43
Total Token throughput (tok/s):          113814.84
---------------Time to First Token----------------
Mean TTFT (ms):                          229.18
Median TTFT (ms):                        231.62
P99 TTFT (ms):                           510.13
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          8.25
Median TPOT (ms):                        8.33
P99 TPOT (ms):                           9.50
---------------Inter-token Latency----------------
Mean ITL (ms):                           8.25
Median ITL (ms):                         6.32
P99 ITL (ms):                            31.12
==================================================
```

### TPx PP Eval


|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.689|±  |0.0146|
|     |       |strict-match    |     5|exact_match|↑  |0.686|±  |0.0147|


### TPx PP Perf 

Perf (Multi Instance)

```
============ Serving Benchmark Result ============
Successful requests:                     512
Maximum request concurrency:             128
Benchmark duration (s):                  7.90
Total input tokens:                      522598
Total generated tokens:                  65536
Request throughput (req/s):              64.77
Output token throughput (tok/s):         8291.08
Total Token throughput (tok/s):          74405.90
---------------Time to First Token----------------
Mean TTFT (ms):                          146.42
Median TTFT (ms):                        123.13
P99 TTFT (ms):                           406.15
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          14.30
Median TPOT (ms):                        14.43
P99 TPOT (ms):                           15.01
---------------Inter-token Latency----------------
Mean ITL (ms):                           14.30
Median ITL (ms):                         12.82
P99 ITL (ms):                            43.65
==================================================
```

Baseline (Single Instance)

```
============ Serving Benchmark Result ============
Successful requests:                     512
Maximum request concurrency:             128
Benchmark duration (s):                  7.76
Total input tokens:                      522598
Total generated tokens:                  65536
Request throughput (req/s):              66.02
Output token throughput (tok/s):         8450.47
Total Token throughput (tok/s):          75836.33
---------------Time to First Token----------------
Mean TTFT (ms):                          144.65
Median TTFT (ms):                        105.95
P99 TTFT (ms):                           406.44
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          14.02
Median TPOT (ms):                        14.20
P99 TPOT (ms):                           14.63
---------------Inter-token Latency----------------
Mean ITL (ms):                           14.02
Median ITL (ms):                         12.54
P99 ITL (ms):                            42.69
==================================================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

luccafong committed:
support non-ray DI

mvoe out dist executor, serving tab

Signed-off-by: Lu Fang <fanglu@fb.com>

formatting

Signed-off-by: Lu Fang <fanglu@fb.com>

format

Signed-off-by: Lu Fang <fanglu@fb.com>

fix lint

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
add tests

Signed-off-by: Lu Fang <fanglu@fb.com>

simon-mo commented:
Is this ready for @njhill to review?

houseroad commented:
Yes, please review.

luccafong committed:
fix lint

Signed-off-by: Lu Fang <fanglu@fb.com>

houseroad reviewed (commented):
One meta question is that if we just extend mp executor, will that be enough?

luccafong committed:
merge mp and mp distributed

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
clean up distributed executor and address comments

Signed-off-by: Lu Fang <fanglu@fb.com>

Signed-off-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

Signed-off-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

Signed-off-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

luccafong commented on vllm/engine/arg_utils.py:
```diff
@@ -705,6 +709,18 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
             **parallel_kwargs["pipeline_parallel_size"],
         )
+        parallel_group.add_argument(
+            "--distributed-master-ip", **parallel_kwargs["distributed_master_ip"]
```
it is included in ParallelConfig.__doc__,

luccafong committed:
add executor test

Signed-off-by: Lucia Fang <fanglu@fb.com>

luccafong committed:
Merge remote-tracking branch 'origin/main' into lucia/non-ray-di

Signed-off-by: Lucia Fang <fanglu@fb.com>

houseroad commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3908,6 +3908,8 @@ def get_attn_backends_for_group(
             # they are cached correctly, there will be different objects per
             # layer.
             for layer_name in kv_cache_group_spec.layer_names:
+                if layer_name not in layers:
```
Shall we add some debugging logging here?

luccafong replied:
@yinghai  it is multi-node PP, where PP rank may not hold all layer_names in the spec

njhill replied:
I'm also curious why this is needed since this class is already used for multi-node PP with ray?

njhill replied:
@luccafong wondering if you missed this question?

houseroad reviewed (commented):
Overall, looks good to me. Some comments to address.

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -65,6 +84,18 @@ def _init_executor(self) -> None:
+        self.local_world_size = (
+            self.parallel_config.world_size
+            // self.parallel_config.distributed_node_size
+        )
```
```suggestion
        self.local_world_size = self.world_size // self.parallel_config.distributed_node_size
```

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -85,40 +116,45 @@ def _init_executor(self) -> None:
+                global_rank = (
+                    self.local_world_size * self.parallel_config.distributed_node_rank
+                    + local_rank
+                )
```
```suggestion
            global_start_rank = self.local_world_size * self.parallel_config.distributed_node_rank
            for local_rank in range(self.local_world_size):
                global_rank = global_start_rank + local_rank
```

luccafong commented:
> > @luccafong thanks for working on this and sorry for taking so long to get to it. A couple of first thoughts from an initial pass:
> > 
> > * Possibly similar to @houseroad's comment, I 'm not sure that we need to modify the core engine or have the concept of a "schedulerless" mode. It would be better to avoid adding complexity to that. Instead I think we should have a separate entry path which runs a "headless" multiproc executor standalone.
> > * In multiproc executor I think we should more cleanly decouple the list of managed procs and list of workers (which will align in the single-node case)
> > * It would be good to have two startup ready checks for each worker. We can change the existing pipe to be duplex and use that for a "local" ready check, and keep the existing messagequeue-based one for the centralized check.
> > * We may need to think about how this works in conjunction with (multi-node) DP
> 
> thanks for the feedbacks and reviews, for first suggestion, do you think if we can reuse headless path used by DP? but we need to adjust some config and naming there.

@njhill  some attempt here https://github.com/luccafong/vllm/commit/ed4acb32a3fd8bba71e4401b2619aa4e13d7da3d

njhill commented:
Thanks @luccafong ...

> do you think if we can reuse headless path used by DP? but we need to adjust some config and naming there.

I'm actually not sure ... we may be able to reuse the topmost part, or to follow that as a reference, but I think other parts will be different since we don't need/want to create engine core (and so would not use `CoreEngineProcManager`); we only need to run a "standalone" multiproc executor.

luccafong commented:
> * We may need to think about how this works in conjunction with (multi-node) DP

I think we can unifiy/merge config for DP config  and the parallel distributed config here (make it generalize) in future PRs

luccafong committed:
exector only

Signed-off-by: Lucia Fang <fanglu@fb.com>

luccafong committed:
revert change to core.py

Signed-off-by: Lucia Fang <fanglu@fb.com>

luccafong committed:
address minor comments

Signed-off-by: Lucia Fang <fanglu@fb.com>

luccafong committed:
compatibility with dp

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
Merge branch 'main' into lucia/non-ray-di

Signed-off-by: Lucia Fang <116399278+luccafong@users.noreply.github.com>

luccafong committed:
fix lint

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
Merge remote-tracking branch 'origin/main' into lucia/non-ray-di

luccafong committed:
add more comments

Signed-off-by: Lu Fang <fanglu@fb.com>

njhill commented on vllm/entrypoints/cli/serve.py:
```diff
@@ -109,6 +106,29 @@ def signal_handler(signum, frame):
     signal.signal(signal.SIGTERM, signal_handler)
     signal.signal(signal.SIGINT, signal_handler)
 
+    if parallel_config.distributed_node_rank_within_dp > 0:
```
Would be good to add a comment here explaining that this is for running "headless workers" mode.

njhill committed:
simplify process

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
bit more cleanup

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
move method

Signed-off-by: Nick Hill <nhill@redhat.com>

luccafong committed:
simply lb mode selection and fix serving and lint issues

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
fix test

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
Merge remote-tracking branch 'origin/main' into lucia/non-ray-di

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
add assertion

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
add back version number in launch msg

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong committed:
add more comments explaining blocking

Signed-off-by: Lu Fang <fanglu@fb.com>

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -234,6 +289,22 @@ def take_draft_token_ids(self) -> DraftTokenIds | None:
+                message_queues.append(remote_message_queue)
+        if unique_reply_rank is not None:
+            message_queues = [message_queues[unique_reply_rank]]
+        return message_queues
```
We should avoid constructing this list each time, can set `self.worker_response_mqs = self._get_response_mqs()` near the end of `_init_executor()` (suggest right before `start_worker_monitor()`. Without unique reply rank of course, that part can stay equivalent to how it already was below...

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -234,6 +289,22 @@ def take_draft_token_ids(self) -> DraftTokenIds | None:
             "take_draft_token_ids", unique_reply_rank=self.output_rank
         )
 
+    def get_response_mqs(
```
suggest rename to indicate private
```suggestion
    def _get_response_mqs(
```

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -410,6 +485,38 @@ class WorkerProc:
 
     READY_STR = "READY"
 
+    def init_message_queues(
```
indicate private
```suggestion
    def _init_message_queues(
```

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -118,41 +131,60 @@ def _init_executor(self) -> None:
-            self.world_size, self.world_size, max_chunk_bytes=max_chunk_bytes
-        )
-        scheduler_output_handle = self.rpc_broadcast_mq.export_handle()
+        self._init_request_rpc_mq()
```
I think it would be a bit clearer to actually inline some of the logic here, like:
```python
            self.rpc_broadcast_mq = None
            self.scheduler_output_handle = None
            if self.parallel_config.distributed_node_rank_within_dp == 0:
                # Only create outbound broadcast queue if we are the leader node
                # within a dp rank, not if we are a headless multiproc executor.
                self.rpc_broadcast_mq = self._init_request_rpc_mq()
                scheduler_output_handle = self.rpc_broadcast_mq.export_handle()
```
and `scheduler_output_handle` doesn't need to be a field.

njhill reviewed (commented):
Thanks @luccafong for all of the great work, I think this is getting very close!

Please see latest comments which are all pretty minor.

The last thing I'm looking / thinking about is the CLI args UX.

youkaichao reviewed (commented):
we can try to align the cli arguments to `torchrun`, like:

`torchrun --nnodes 2 --nproc-per-node=2 --node-rank 0 --master_addr $MASTER_ADDR test.py`

we can use `--nnodes` `--node-rank` `--master_addr` .
`torchrun` has a broad audience and this reduces the pain for people to learn.

luccafong committed:
arg change and model executor cleanup

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong commented:
> we can try to align the cli arguments to `torchrun`, like:
> 
> `torchrun --nnodes 2 --nproc-per-node=2 --node-rank 0 --master_addr $MASTER_ADDR test.py`
> 
> we can use `--nnodes` `--node-rank` `--master_addr` . `torchrun` has a broad audience and this reduces the pain for people to learn.

thanks for the suggestions, I adjusted the args accordingly and updated the PR summary

luccafong committed:
clean up gpu model runner code that is not needed

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong commented:
@njhill  for comment https://github.com/vllm-project/vllm/pull/23691#discussion_r2520529547

I found this now no longer an issue without these 2 lines, so cleaned them up, tested on both pp*tp and dp*pp*tp, working well.

luccafong committed:
fix lint

Signed-off-by: Lu Fang <fanglu@fb.com>

luccafong commented:
@njhill @youkaichao  I have addressed comments and args, retested all combinations on lm_eval.

luccafong committed:
remove print

Signed-off-by: Lu Fang <fanglu@fb.com>

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -144,15 +165,36 @@ def _init_executor(self) -> None:
 
+            # Start background thread to monitor worker health if not in headless mode.
+            if self.monitor_workers:
+                self.start_worker_monitor()
```
nit: add empty line here?

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -234,6 +280,22 @@ def take_draft_token_ids(self) -> DraftTokenIds | None:
+                message_queues.append(remote_message_queue)
+        if unique_reply_rank is not None:
+            message_queues = [message_queues[unique_reply_rank]]
+        return message_queues
```
I think you forgot to delete this method?

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -268,20 +332,20 @@ def collective_rpc(  # type: ignore[override]
-        )
+        response_mqs = self.response_mqs
+        if unique_reply_rank is not None:
+            response_mqs = [self.response_mqs[unique_reply_rank]]
```
Use `output_rank` here, also cheaper to use singleton tuple.
```suggestion
        response_mqs: Sequence[MessageQueue] = self.response_mqs
        if output_rank is not None:
            response_mqs = (response_mqs[output_rank],)
```

njhill reviewed (commented):
Thanks @luccafong. A few small things...

luccafong committed:
address commment

Signed-off-by: Lu Fang <fanglu@fb.com>

houseroad reviewed (approved):
Offline synced with @youkaichao  and @njhill, and aligned.

Let's land this for now.

luccafong committed:
fix test for the positional arg used

Signed-off-by: Lu Fang <fanglu@fb.com>

jianzs commented:
@luccafong  Great work on this!

I have a question regarding the design: building on these changes, do you think it is feasible to completely abstract away or remove the node concept from the mp_executor?

I am currently implementing scale up/down functionality for the MP backend. However, I've found that the explicit node concept restricts flexibility when resizing. specifically, if I want to scale down just a single DP rank within a node, it breaks the consistency of `nnodes_within_dp` across different nodes. I am concerned this inconsistency might lead to exceptions or undefined behavior.

Would love to hear your thoughts on how to handle this granular scaling.
